### PR TITLE
Optimizing code and adding from http to https

### PR DESCRIPTION
**Pull request by rajendraarora16**

### CODE REVIEW

File: tensorflow/c/c_api.cc [hunk 1]
```diff
@@ -262,7 +262,7 @@ static Status TF_StringDecode_Impl(const char* src, size_t src_len,
   if (p == nullptr) {
     return InvalidArgument("invalid string encoding or truncated src buffer");
   }
-  if (len64 > std::numeric_limits<size_t>::max()) {
+  else if (len64 > std::numeric_limits<size_t>::max()) {
```
Comment (asimshankar) [inline tensorflow/c/c_api.cc]: Why this change? The previous `if` block contains an unconditional `return`, so why does the `else` here help?

File: tensorflow/c/eager/tape.h [hunk 2]
```diff
@@ -216,23 +216,17 @@ void GradientTape<Gradient, BackwardFunction>::RecordOperation(
...
-  tensor_usage_.erase(it);
   auto tensor_op_it = tensor_tape_.find(tensor_id);
-  if (tensor_op_it == tensor_tape_.end()) {
-    return;
-  }
   const int64 op_id = tensor_op_it->second;
-  if (op_id == -1) {
-    // Do not delete watched tensors.
+
+  it->second--;
```
Comment (asimshankar) [inline tensorflow/c/eager/tape.h]: I didn't quite follow the change here. What is the "optimization" - is it lines of code or some performance argument being made here?

**Review by asimshankar — COMMENTED** [review]
Thanks for the changes, but I must admit I don't quite follow the motivation behind or the benefits of most of them? Could you elaborate?

If not, reducing this to the changes in `while_op.cc` and `faq.md` seems reasonable to me.

File: tensorflow/c/eager/tape.h (hunk 2 above)
Comment (rajendraarora16) [reply tensorflow/c/eager/tape.h]: Thanks for reviewing my changes, 
Just have found multiple times if-statement was being used to exit function that's why decided to reduce the lines of code and making it with single statement to pass `empty return`. Just have followed the same single statement as it is done in `line no 322`. If you think it is not valid. I will revert it back. Just wanted to reduce the lines of code.

Comment (asimshankar) [reply tensorflow/c/eager/tape.h]: I'd suggest we leave this as is. I know this is subjective, but I don't think the 4 lines saved improve readability of the logic (and it doesn't introduce questions like is it okay to use `it->second--` when there was no entry in `tensor_usage_` etc.

**Review by asimshankar — CHANGES_REQUESTED** [review]

File: tensorflow/c/eager/tape.h (hunk 2 above)
Comment (rajendraarora16) [reply tensorflow/c/eager/tape.h]: Thanks! Changes are reverted as per your suggestion. Can we go ahead :)

**Comment by rajendraarora16** [comment]
Thanks! Changes are done as per your suggestion. Can we go ahead :)

**Review by asimshankar — APPROVED** [review]

**Comment by drpngx** [comment]
Jenkins, test this please.
